### PR TITLE
feat: remove deploy from directory

### PR DESCRIPTION
**Pull request by SimonRichardson**
Removes the ability to deploy/refresh from a directory. Instead,
Juju now expects the charm to be in the form of a zip archive.

This aligns Juju with `charmcraft pack` output.

---- 

If you can't archive a folder locally, we no longer need to have
the ability to generate a version file. We will still read the
version file in the zip file if it's present, but we won't
generate it. Along with that, the `.jujuignore` file is also gone.
The same principle applies, that was only ever-present because
of how we archived files. If we're not creating archives then we
no longer need it.

By simplifying the deployment process, we're also reducing the
complexity of a charm. We don't need a .jujuignore file, thus we
don't potentially need a version file. That information could be
stored in the `manifest.yaml`.

Note: everything that applies to charms, also applies to bundles
deploying local charm directories. Charms must be prepackaged;
ideally with `charmcraft pack`.

We're also dropping dependencies, as we no longer globbing for files
when attempting to archive directories.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

### Deploy charm

```sh
$ juju bootstrap lxd test
$ juju add-model test

$ cd testcharms/charm-hub/charms/juju-qa-test
$ charmcraft pack
$ juju deploy ./juju-qa-test_ubuntu-22.04-amd64-arm64_ubuntu-24.04-amd64-arm64.charm
```

### Deploy bundle

It expects that all bundles will have paths to archives.

```yaml
applications:
  juju-qa-test:
    charm: <path/to/charm>.charm
    num_units: 1
```

### Refresh

```sh
$ cd testcharms/charm-hub/charms/juju-qa-test
$ charmcraft pack
$ juju refresh --path ./juju-qa-test_ubuntu-22.04-amd64-arm64_ubuntu-24.04-amd64-arm64.charm
```

### Local deployment

.charm files are just .zip files with a different extension.
To regain the same deploy from directory workflow, you
can utilize:

```sh
$ zip -r <name>.charm <dir>
$ juju deploy <name>.charm
``` 

To produce the same iterative workflow as a directory, zip files can
be amended with the updated file, so a workflow could be:

```sh
$ zip <name>.charm <path/to/file>
$ juju deploy <name>.charm
```

## Documentation changes

@tmihoc Removal of deploy from directory.

## Links

**Jira card:** JUJU-

### CODE REVIEW

File: cmd/juju/application/deploy.go [hunk 1]
```diff
@@ -364,16 +364,11 @@ If a revision is specified, a channel must also be specified for Charmhub charms
...
-
-  juju deploy /path/to/charm
-  juju deploy /path/to/charm --base ubuntu@22.04
-
 You will need to be explicit if there is an ambiguity between a local and a
 remote charm:
 
-  juju deploy ./pig
-  juju deploy ch:pig
+  juju deploy ./postgresql
```
Comment (hmlanigan) [inline cmd/juju/application/deploy.go:370]: `juju deploy ./postgreql.charm` helps to emphasize charm craft packed charm?

File: cmd/juju/application/deploy.go [hunk 2]
```diff
@@ -364,16 +364,11 @@ If a revision is specified, a channel must also be specified for Charmhub charms
 and bundles.  The charm will be deployed with revision.  The channel will be used
 when refreshing the application in the future.
 
-A local charm may be deployed by giving the path to its directory:
-
-  juju deploy /path/to/charm
-  juju deploy /path/to/charm --base ubuntu@22.04
-
```
Comment (hmlanigan) [inline cmd/juju/application/deploy.go:371]: We should elaborate on what the local charm is to be clear. `charmcraft pack` uses the term `charm operator package`?

```
A local charm may be deploy by giving the path to the charm package:

   juju deploy /path/to/example.charm

A charm package is created with charmcraft pack command, or by creating a zip file.
```

File: cmd/juju/application/refresh.go [hunk 3]
```diff
@@ -266,8 +266,8 @@ func (c *refreshCommand) SetFlags(f *gnuflag.FlagSet) {
 	f.StringVar(&c.SwitchURL, "switch", "", "Crossgrade to a different charm")
 	f.StringVar(&c.CharmPath, "path", "", "Refresh to a charm located at path")
```
Comment (hmlanigan) [inline cmd/juju/application/refresh.go:267]: 'a charm package' to disambiguate.

File: cmd/juju/application/refresh.go [hunk 4]
```diff
@@ -174,8 +174,8 @@ uploaded with the revision specified in the charm, if possible, otherwise it
 gets a unique revision (highest in state + 1).
 
 When deploying from a path, the --path option is used to specify the location from
-which to load the updated charm. Note that the directory containing the charm must
-match what was originally used to deploy the charm as a superficial check that the
+which to load the updated charm. Note that the charm must match what was 
+originally used to deploy the charm as a superficial check that the
```
Comment (hmlanigan) [inline cmd/juju/application/refresh.go:178]: its not clear what form the charm is in here. Be specific about the location of the charm package. Or whichever term we use.

Comment (hmlanigan) [reply cmd/juju/application/refresh.go:178]: There is a reference to local charm paths at line 167 as well.

**Review by hmlanigan — COMMENTED** [review]
The change looks good. Comments about wording in the help/error messages and CI tests which will start failing.

It looks like multiple CI tests will start failing:

- storage with ./testcharms/charms/dummy-storage*
- relations with ./testcharms/charms/departer, /testcharms/charms/dummy-sink, /testcharms/charms/dummy-source
- deploy with ./testcharms/charms/lxd-profile, ./testcharms/charms/lxd-profile-subordinate

Maybe more.

The dummy source and sink charms do have Charmhub versions, not sure of the difference with the local charms.

**Comment by SimonRichardson** [comment]
I've assigned @jack-w-shaw to go through the deploy CI tests in main because a lot are broken with his change to require the manifest.yaml.

File: cmd/juju/application/deploy.go (hunk 1 above)
Comment (SimonRichardson) [reply cmd/juju/application/deploy.go:370]: Agreed.

File: cmd/juju/application/deploy.go (hunk 2 above)
Comment (SimonRichardson) [reply cmd/juju/application/deploy.go:371]: Agreed, but I'm dropping `, or by creating a zip file.` to illustrate that `charmcraft pack` is the recommended way. We could add developer documentation to state what makes a charm file. The help docs isn't the place for that.

**Review by hmlanigan — APPROVED** [review]
LGTM.

**Pull request merged**